lib/puppet/provider/postgresql_conf/ruby.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+Puppet::Type.type(:postgresql_conf).provide(:ruby) do
+  desc 'Set key/values in a postgresql config file.'
+  confine kernel: 'Linux'
+
+  # file parsing in a seperate function
+  def parse_config
+    file = File.open(resource[:target])
+    active_values_regex = %r{^\s*(?<key>[\w.]+)\s*=?\s*(?<value>.*?)(?:\s*#\s*(?<comment>.*))?\s*$}
+    active_settings = []
+
+    File.foreach(file).with_index do |line, index|
+      line_number = index + 1
+      matches = line.match(active_values_regex)
+      if matches
+        attributes_hash = { line_number: line_number, key: matches[:key], ensure: 'present', value: matches[:value], comment: matches[:comment] }
+        active_settings.push(attributes_hash)
+      end
+    end
+    active_settings
+  end
+
+  # delete existing header
+  def delete_header(lines)
+    header_regex = %r{^# HEADER:.*}
+    lines.delete_if do |entry|
+      entry.match?(header_regex)
+    end
+  end
+
+  def add_header(lines)
+    timestamp = Time.now.strftime('%F %T %z')
+    header = ["# HEADER: This file was autogenerated at #{timestamp}\n",
```

Comment:
Will the timestamp be changed on every puppet run?

Reply:
the write_config function is called on create and delete operations and deletes the existing header (if any), adds a new one with up to date timestamp and writes the changes to file. As far as I tested this right now, this works reliable.